File: CSharp/Yow.CoD.Finance.Domain/Contracts/CreateLoanCommand.cs
using System;

namespace Yow.CoD.Finance.Domain.Contracts
{
    public class CreateLoanCommand : Command
    {
        public CreateLoanCommand(Guid commandId,
            Guid aggregateId,
            DateTimeOffset createdOn,
            CustomerContact customerContact,
            BankAccount bankAccount,
            PaymentPlan paymentPlan,
            decimal amount,
            Duration term)
            : base(commandId, aggregateId)
        {
            if (createdOn == default(DateTimeOffset)) throw new ArgumentException("CreatedOn must be non default value", nameof(createdOn));
            if (customerContact == null) throw new ArgumentException("CustomerContact must be non null", nameof(customerContact));
            if (bankAccount == null) throw new ArgumentException("BankAccount must be non null", nameof(bankAccount));
            if (paymentPlan == PaymentPlan.None) throw new ArgumentException("PaymentPlan must be non default value", nameof(paymentPlan));
            if (amount == 0) throw new ArgumentException("Amount must be non zero value", nameof(amount));
            if (term == null) throw new ArgumentException("Term must be non null", nameof(term));

            CreatedOn = createdOn;
            CustomerContact = customerContact;
            BankAccount = bankAccount;
            PaymentPlan = paymentPlan;
            Amount = amount;
            Term = term;
        }

        public DateTimeOffset CreatedOn { get; }

        public CustomerContact CustomerContact { get; }

        public BankAccount BankAccount { get; }

        public PaymentPlan PaymentPlan { get; }

        public decimal Amount { get; }

        public Duration Term { get; }

        public override string ToString()
        {
            return $"CreateLoanCommand{{ CreatedOn:'{CreatedOn}', Amount:'{Amount}', Term:{Term}}}";
        }
    }
}

File: typescript/packages/domain/src/contracts/create-loan-command.ts
import { InvalidCommandError } from "../errors.js";
import { BankAccount } from "./bank-account.js";
import { Command } from "./command.js";
import { CustomerContact } from "./customer-contact.js";
import { Duration } from "./duration.js";
import { PaymentPlan } from "./payment-plan.js";

export class CreateLoanCommand extends Command {
  readonly #brand!: undefined;
  readonly createdOn: Date;
  readonly customerContact: CustomerContact;
  readonly bankAccount: BankAccount;
  readonly paymentPlan: PaymentPlan;
  readonly amount: number;
  readonly term: Duration;

  private constructor(
    commandId: string,
    aggregateId: string,
    createdOn: Date,
    customerContact: CustomerContact,
    bankAccount: BankAccount,
    paymentPlan: PaymentPlan,
    amount: number,
    term: Duration,
  ) {
    super(commandId, aggregateId);
    this.createdOn = createdOn;
    this.customerContact = customerContact;
    this.bankAccount = bankAccount;
    this.paymentPlan = paymentPlan;
    this.amount = amount;
    this.term = term;
  }

  static create(input: {
    commandId: string;
    aggregateId: string;
    createdOn: Date;
    customerContact: CustomerContact;
    bankAccount: BankAccount;
    paymentPlan: PaymentPlan;
    amount: number;
    term: Duration;
  }): CreateLoanCommand {
    if (!(input.createdOn instanceof Date) || isNaN(input.createdOn.getTime()))
      throw new InvalidCommandError("CreatedOn must be a valid date");
    if (!input.customerContact)
      throw new InvalidCommandError("CustomerContact must be non null");
    if (!input.bankAccount)
      throw new InvalidCommandError("BankAccount must be non null");
    if (!input.paymentPlan)
      throw new InvalidCommandError("PaymentPlan must be non default value");
    if (input.amount === 0)
      throw new InvalidCommandError("Amount must be non zero value");
    if (!input.term)
      throw new InvalidCommandError("Term must be non null");

    return new CreateLoanCommand(
      input.commandId,
      input.aggregateId,
      input.createdOn,
      input.customerContact,
      input.bankAccount,
      input.paymentPlan,
      input.amount,
      input.term,
    );
  }

  toString(): string {
    return `CreateLoanCommand{ CreatedOn:'${this.createdOn.toISOString()}', Amount:'${this.amount}', Term:${this.term}}`;
  }
}
